### PR TITLE
Add URL for downloading CLI to the console-config configmap

### PR DESCRIPTION
**Pull request by jhadvig**
PR for adding the URL for downloading CLI to the `console-config` configmap, in `clusterInfo.cliDownloadURL` field.

Screen Shot 2019-09-25 at 21 27 02

Story: https://jira.coreos.com/browse/CONSOLE-1423

/assign @benjaminapetersen

### CODE REVIEW

**Comment by spadgett** [comment]
We'll a way to specify different URLs for different platforms

**Comment by jhadvig** [comment]
@spadgett the downloads pod contains binary for linux,osx, widnows... was in impression that it is the route to the pod that we want to wire to the console.

**Comment by jhadvig** [comment]
Or we want a direct link for each platform ?

**Comment by spadgett** [comment]
> Or we want a direct link for each platform in the console's "Command Line Tools" page ?

Yes, IMO this is a much better experience than dumping the user in a directory listing.

**Comment by jhadvig** [comment]
@spadgett also, for each platform there are is a: `oc`, `oc.tar`, `oc.zip`
I suppose we want to provide the plane `oc` binary since its a single file ?

**Comment by spadgett** [comment]
Hm, maybe linux .tar.gz and mac and windows use .zip ?

**Comment by jhadvig** [comment]
@spadgett @benjaminapetersen  I've update the PR
Now there will be a URL for each platform we build binary for.

PTAL

Screen Shot 2019-09-26 at 17 34 35

File: pkg/console/subresource/configmap/configmap.go [hunk 1]
```diff
@@ -39,14 +46,16 @@ func DefaultConfigMap(
 	consoleConfig *configv1.Console,
 	managedConfig *corev1.ConfigMap,
 	infrastructureConfig *configv1.Infrastructure,
-	rt *routev1.Route) (consoleConfigmap *corev1.ConfigMap, unsupportedOverridesHaveMerged bool, err error) {
+	rt *routev1.Route,
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap.go]: Can probably update to 

```golang
consoleRoute *routev1.Route,
downloadsRoute *routev1.Route) (consoleConfigmap *corev1.ConfigMap, unsupportedOverridesHaveMerged bool, err error) {
```
Mean name both routes now that there are two.

File: pkg/console/subresource/configmap/configmap_test.go [hunk 2]
```diff
@@ -19,6 +19,10 @@ import (
 
 const (
 	host               = "localhost"
+	downloadHost       = "https://downloads-openshift-console.apps.some.cluster.openshift.com"
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap_test.go]: To be honest, for the test I'd rather inline these.  Its a tad repetitious, but I think it makes a better test if the `input` and `output` are specific to each test and the `output` (ie, the want) is hard-coded in the specific test.

Comment (benjaminapetersen) [reply pkg/console/subresource/configmap/configmap_test.go]: It may be good to add one test w/o the downloads route (`nil`) and make sure the config is still properly generated:

```golang
{
     name: "Test default configmap, no customization overrides",
     // downloadsRoute: &routev1.Route{ ... }
}, {
     name: "Test default configmap, without downloads",
     // downloadsRoute: nil
}
```

File: pkg/console/subresource/consoleserver/types.go [hunk 3]
```diff
@@ -39,9 +39,17 @@ type ServingInfo struct {
...
-	ConsoleBasePath    string `yaml:"consoleBasePath,omitempty"`
-	MasterPublicURL    string `yaml:"masterPublicURL,omitempty"`
+	ConsoleBaseAddress string           `yaml:"consoleBaseAddress,omitempty"`
+	ConsoleBasePath    string           `yaml:"consoleBasePath,omitempty"`
+	MasterPublicURL    string           `yaml:"masterPublicURL,omitempty"`
+	CLIDownloadURLs    *CLIDownloadURLs `yaml:"cliDownloadURLs,omitempty"`
+}
+
+// CLIDownloadURLs contains download URLs for each of the platforms we provide CLI binary.
+type CLIDownloadURLs struct {
```
Comment (benjaminapetersen) [inline pkg/console/subresource/consoleserver/types.go]: Maybe simplify this a bit:

```golang
type CLIDownloadURLs struct {
	Linux   string `yaml:"linuxDownloadURL"`
	Mac    string `yaml:"macDownloadURL"`
	Windows string `yaml:"windowsDownloadURL"`
```
Since its already a heading `CLIDownloadURLs` we don't need to repeat the `DownloadURL` part in the field names.

File: pkg/console/subresource/util/util.go [hunk 4]
```diff
@@ -106,3 +106,9 @@ func HTTPS(host string) string {
 	secured := fmt.Sprintf("%s%s", protocol, host)
 	return secured
 }
+
```
Comment (benjaminapetersen) [inline pkg/console/subresource/util/util.go]: I think you can keep this func in the `configmap_builder.go` file.  It isn't shared elsewhere.

Comment (benjaminapetersen) [reply pkg/console/subresource/util/util.go]: Also, should we give it a unit test?

File: pkg/api/api.go [hunk 5]
```diff
@@ -1,5 +1,12 @@
 package api
 
+const (
```
Comment (benjaminapetersen) [inline pkg/api/api.go]: I'm inclined to move these into the `configmap_builder.go` file, since they are only used in one place.

File: pkg/console/operator/sync_v400.go [hunk 6]
```diff
@@ -316,7 +316,12 @@ func (co *consoleOperator) SyncConfigMap(
 		return nil, false, "FailedManagedConfig", mcErr
 	}
 
-	defaultConfigmap, _, err := configmapsub.DefaultConfigMap(operatorConfig, consoleConfig, managedConfig, infrastructureConfig, rt)
+	downloadsRoute, err := co.routeClient.Routes(api.TargetNamespace).Get("downloads", metav1.GetOptions{})
+	if err != nil {
+		return nil, false, "FailedGetDownloadsRoute", err
+	}
+
+	defaultConfigmap, _, err := configmapsub.DefaultConfigMap(operatorConfig, consoleConfig, managedConfig, infrastructureConfig, rt, downloadsRoute)
```
Comment (benjaminapetersen) [inline pkg/console/operator/sync_v400.go]: want to update the name to `consoleRoute, downloadsRoute`?

File: pkg/console/subresource/configmap/configmap_test.go [hunk 7]
```diff
@@ -332,14 +382,137 @@ servingInfo:
...
+  certFile: /var/serving-cert/tls.crt
+  keyFile: /var/serving-cert/tls.key
+providers: 
+  statuspageID: id-1234
+`,
+				},
+			},
+		},
+		{
+			name: "Test openshift-console namespace without downloads route",
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap_test.go]: +1 for a test `without` the downloads route.
`Test openshift-console namespace without downloads route` - `namespace`?  I'm a bit confused by the test name, however.

File: pkg/console/subresource/configmap/configmap_test.go [hunk 8]
```diff
@@ -332,14 +382,137 @@ servingInfo:
   keyFile: /var/serving-cert/tls.key
 providers: 
   statuspageID: id-1234
+`,
+				},
+			},
+		},
+		{
+			name: "Test operator config with Statuspage pageID",
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap_test.go]: I believe we already had this test, trying to figure out why its showing up as new.

File: pkg/console/subresource/consoleserver/config_builder.go [hunk 9]
```diff
@@ -58,6 +59,18 @@ func (b *ConsoleServerCLIConfigBuilder) APIServerURL(apiServerURL string) *Conso
 	b.apiServerURL = apiServerURL
 	return b
 }
+func (b *ConsoleServerCLIConfigBuilder) CLIDownloadURL(cliDownloadURL string) *ConsoleServerCLIConfigBuilder {
+	if cliDownloadURL != "" {
+		b.cliDownloadURLs = &CLIDownloadURLs{
```
Comment (benjaminapetersen) [inline pkg/console/subresource/consoleserver/config_builder.go]: See my other comment about dropping the `DownloadURL` redundancy. I think we can clean this up just a bit.

File: pkg/console/subresource/consoleserver/types.go (hunk 3 above)
Comment (benjaminapetersen) [reply pkg/console/subresource/consoleserver/types.go]: To be honest, we may want to do CLIDownloadURLs[arch][platform] just like the downloads deployment. Only amd64 now.

File: pkg/console/subresource/util/util.go (hunk 4 above)
Comment (jhadvig) [reply pkg/console/subresource/util/util.go]: will add !

File: pkg/console/subresource/configmap/configmap_test.go (hunk 7 above)
Comment (jhadvig) [reply pkg/console/subresource/configmap/configmap_test.go]: Wasnt really sure how to name the test, other then this since the setup is `nil` download route. But really should be something like : `Test getting URLs for downloading CLI, without downloads route.` ... maybe ?

File: pkg/console/subresource/consoleserver/config_builder.go (hunk 9 above)
Comment (jhadvig) [reply pkg/console/subresource/consoleserver/config_builder.go]: makes sense !

File: pkg/console/subresource/configmap/configmap_test.go (hunk 2 above)
Comment (jhadvig) [reply pkg/console/subresource/configmap/configmap_test.go]: > It may be good to add one test w/o the downloads route (nil) and make sure the config is still properly generated:

All other TCs have the download route set :)

File: pkg/console/subresource/consoleserver/types.go (hunk 3 above)
Comment (jhadvig) [reply pkg/console/subresource/consoleserver/types.go]: @benjaminapetersen how about? 

```go
type ClusterInfo struct {
	ConsoleBaseAddress string          `yaml:"consoleBaseAddress,omitempty"`
	ConsoleBasePath    string          `yaml:"consoleBasePath,omitempty"`
	MasterPublicURL    string          `yaml:"masterPublicURL,omitempty"`
	CLIDownloadURLs    CLIDownloadURLs `yaml:"cliDownloadURLs,omitempty"`
}

// CLIDownloadURLs contains all the architertures that we are building CLI binaries for.
type CLIDownloadURLs struct {
	Amd64 ArchPlatformsURL `yaml:"amd64,omitempty"`
}

// ArchPlatformsURL contains URLs for each of the platforms we provide CLI binary.
type ArchPlatformsURL struct {
	Linux   string `yaml:"linux"`
	Mac     string `yaml:"mac"`
	Windows string `yaml:"windows"`
}
```

Although Im not really sure about the `CLIDownloadURLs`, but its pretty generic and with the aditional architecture specified it kinda dioes make sense?

File: pkg/console/subresource/configmap/configmap_test.go (hunk 7 above)
Comment (benjaminapetersen) [reply pkg/console/subresource/configmap/configmap_test.go]: The expectation is that there would be no downloads then, if there is no route, correct?  I'm good with long tests names that clearly indicate everything, including expected output.

Perhaps `Validate that there is no download CLI links when download route is `nil``?

Out test naming convention isn't super strong, so if we have to break it for clarity, thats fine.  We could always go back and improve the names of the other tests if you have something that works better.

File: pkg/console/subresource/consoleserver/types.go (hunk 3 above)
Comment (benjaminapetersen) [reply pkg/console/subresource/consoleserver/types.go]: Yup!  Though I wonder if `ArchPlatformsURL` vs `ArchPlatformURLs`.  As "one platform has many urls", and if you had many archs, you would `[]ArchPlatformURLs`,

```golang
type CLIDownloadURLs struct {
   Amd64 ArchPlatformURLs `yaml:"amd64,omitempty"`
   Foo32 ArchPlatformURLs `yaml:"Foo32,omitempty"`
   Bar64 ArchPlatformURLs `yaml:"Bar64,omitempty"`
}
```
Since we expect 3+ URLs for archs added.

Comment (jhadvig) [reply pkg/console/subresource/consoleserver/types.go]: But if we tan to chain it like you mentioned in your previous comment it cant be an array.

Comment (jhadvig) [reply pkg/console/subresource/consoleserver/types.go]: +1 on `ArchPlatformURLs`

File: pkg/console/subresource/configmap/configmap_test.go (hunk 2 above)
Comment (jhadvig) [reply pkg/console/subresource/configmap/configmap_test.go]: Also the we have the `host` const here

**Comment by jhadvig** [comment]
@benjaminapetersen comments addressed. PTAL

**Comment by jhadvig** [comment]
@benjaminapetersen so was looking into the console side of this story and figured out that it might be better to create [ConsoleCLIDownload](https://github.com/openshift/console-operator/blob/master/manifests/00-crd-extension-console-cli-download.yaml) CR that will represent the per architecture binaries for all the different platforms.

Eg. I've created following CR :

```
yaml
apiVersion: console.openshift.io/v1
kind: ConsoleCLIDownload
metadata:
  name: download-cli-amd64
spec:
  description: >
    With the OpenShift command line interface, you can create applications and
    manage OpenShift projects from a terminal.


    The oc binary offers the same capabilities as the kubectl binary, but it is
    further extended to natively support OpenShift Container Platform features.
  displayName: oc - OpenShift Command Line Interface (CLI)  - amd64
  links:
    - href: 'https://www.example.com'
      text: Download oc for Linux
    - href: 'https://www.example.com'
      text: Download oc for Mac
    - href: 'https://www.example.com'
      text: Download oc for Windows
```

which will produce the third section in the below screenshot:

Screen Shot 2019-09-30 at 13 02 42


With this I was thinking that instead of adding logic to operator and console, let the operator create the CR for each arch type, that will contain the binaries for all the platforms.

cc'ing @spadgett for opinion

**Comment by benjaminapetersen** [comment]
Ha, I like this. leverages what we already built, no point in all the additional wiring.

File: pkg/console/subresource/configmap/configmap.go [hunk 10]
```diff
@@ -27,6 +27,13 @@ func getApiUrl(infrastructureConfig *configv1.Infrastructure) string {
 	return ""
 }
 
+func getCliDownloadUrl(downloadsRoute *routev1.Route) string {
+	if downloadsRoute != nil {
+		return downloadsRoute.Spec.Host
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap.go]: `route.Spec.Host` will actually be removed, we have a `GetCanonicalHost()` func that could be used instead.

**Comment by spadgett** [comment]
If we're going to do this for `oc`, we might want to do it for `odo` as well. I'd suggest special handling to make sure `oc` is moved to the top of the list. We'll need to handle the case where you are running against native k8s and this resource doesn't exist.

Only potential downside I see is we can't give special handling to the different platforms (e.g., highlight the macos link if we detect you are on a mac).

**Comment by benjaminapetersen** [comment]
@spadgett  @jhadvig agree we discussed the ordering problem a little bit.  `ConsoleCLIDownload` as a type is pretty specific, We could add an optional `platform` tag to each link, which would give us the opportunity to have the filtering logic in the `console`.

**Comment by benjaminapetersen** [comment]
Something like this could be worth discussing:
```yaml
apiVersion: console.openshift.io/v1
kind: ConsoleCLIDownload
metadata:
  name: example-oc
spec:
  displayName: example-oc
  description: >
    This is a link with platforms
  links:
    - href: 'https://www.example.com'
      name: oc for linux
      platform: linux
    - href: 'https://www.example.com'
      name: oc for mac
      platform: mac
    - href: 'https://www.example.com'
      name: oc for windows
      platform: windows
```

**Comment by benjaminapetersen** [comment]
/hold

As we investigate this other direction.

**Comment by jhadvig** [comment]
Closing in favour of https://github.com/openshift/console-operator/pull/306
